src/btf/compat.cpp

```diff
@@ -0,0 +1,116 @@
+#include "btf/compat.h"
+#include "struct.h"
+
+namespace bpftrace::btf {
+
+char CompatTypeError::ID;
+
+void CompatTypeError::log(llvm::raw_ostream &OS) const
+{
+  OS << msg_;
+}
+
+Result<SizedType> getCompatType([[maybe_unused]] const Void &type)
+{
+  return CreateVoid();
+}
+
+Result<SizedType> getCompatType(const Integer &type)
+{
+  return CreateInt(8 * type.bytes());
+}
+
+Result<SizedType> getCompatType(const Pointer &type)
+{
+  auto t = type.element_type();
+  if (!t) {
+    return t.takeError();
+  }
+  auto ty = getCompatType(*t);
+  if (!ty) {
+    return ty.takeError();
+  }
+  return CreatePointer(*ty);
+}
+
+Result<SizedType> getCompatType(const Array &type)
+{
+  auto t = type.element_type();
+  if (!t) {
+    return t.takeError();
+  }
+  auto ty = getCompatType(*t);
+  if (!ty) {
+    return ty.takeError();
+  }
+  return CreateArray(type.element_count(), *ty);
+}
+
+Result<SizedType> getCompatType(const Struct &type)
+{
+  std::vector<SizedType> fields;
+  std::vector<std::string_view> idents;
+  auto f = type.fields();
+  if (!f) {
+    return f.takeError();
+  }
+  // We have no ability to control how things are packed,
+  // so this may or may not generate the correct type.
+  for (const auto &[name, info] : *f) {
+    auto ft = getCompatType(info.type);
+    if (!ft) {
+      return ft.takeError();
+    }
+    fields.emplace_back(*ft);
+    idents.emplace_back(name);
+  }
+  return CreateRecord(bpftrace::Struct::CreateRecord(fields, idents));
+}
+
+Result<SizedType> getCompatType(const Enum &type)
+{
+  auto size = type.size();
+  if (!size) {
+    return size.takeError();
+  }
+  return CreateEnum(8 * (*size), type.name());
+}
+
+Result<SizedType> getCompatType(const Enum64 &type)
+{
+  auto size = type.size();
+  if (!size) {
+    return size.takeError();
+  }
+  return CreateEnum(8 * (*size), type.name());
```

Comment:
Do we need to get the size here? Don't we know it's a 64 bit type?

Reply:
Maybe? I don't think it's any less correct to just get the size. The enum64 is defined as capable of holding 64-bit values and it should always return 8 here, but I'd rather just rely on that underlying BTF call then hard-code something and outsmart it :)

Reply:
> I don't think it's any less correct to just get the size

Sure. I would just lose my mind if the size was ever something other than 8 for enum64